File: clients/Ruby/debug_interface.rb
require_relative 'model'

class DebugInterface
    def initialize(reader, writer)
        @reader = reader
        @writer = writer
    end

    def send(command)
        ClientMessage::DebugMessage.new(command).write_to(@writer)
        @writer.flush()
    end

    def get_state()
        ClientMessage::RequestDebugState.new().write_to(@writer)
        @writer.flush()
        return DebugState.read_from(@reader)
    end
end

File: clients/TypeScript/src/DebugInterface.ts
import {ClientMessage} from './model/ClientMessage';
import {DebugCommand} from './model/DebugCommand';
import {DebugState} from './model/DebugState';
import {StreamWrapper} from './StreamWrapper';

export class DebugInterface {
    constructor(
        private streamWrapper: StreamWrapper,
    ) {}

    async send(command: DebugCommand) {
        await (new ClientMessage.DebugMessage(command)).writeTo(this.streamWrapper);
        // TODO: only flush stream once here?
    }

    async getState() {
        await (new ClientMessage.RequestDebugState()).writeTo(this.streamWrapper);
        // TODO: only flush stream once here?
        return await DebugState.readFrom(this.streamWrapper);
    }
}
